[LgTv] Exception upon startup: "The deactivate method has thrown an exception"

## Expected Behavior
The binding loads and works

## Current Behavior
When I'm booting up, I get the follwing JAVA exception:
`2017-12-08 20:42:29.936 [INFO ] [binding.lgtv.internal.LgtvConnection] - Stopped LgTv Servlet 

2017-12-08 20:42:29.944 [ERROR] [org.openhab.binding.lgtv            ] - [org.openhab.binding.lgtv(207)] The deactivate method has thrown an exception
java.lang.IllegalArgumentException: Alias [/udap/api/event] was never registered
	at org.ops4j.pax.web.service.internal.HttpServiceStarted.unregister(HttpServiceStarted.java:313) ~[?:?]
	at org.ops4j.pax.web.service.internal.HttpServiceProxy.unregister(HttpServiceProxy.java:78) ~[?:?]
	at org.openhab.binding.lgtv.lginteraction.LgTvMessageReader.deactivate(LgTvMessageReader.java:178) ~[?:?]
at org.openhab.binding.lgtv.lginteraction.LgTvMessageReader.stopserver(LgTvMessageReader.java:195) ~[?:?]
at org.openhab.binding.lgtv.internal.LgtvConnection.closeConnection(LgtvConnection.java:158) ~[?:?]
at org.openhab.binding.lgtv.internal.LgtvBinding.closeAllConnections(LgtvBinding.java:378) ~[?:?]
at org.openhab.binding.lgtv.internal.LgtvBinding.deactivate(LgtvBinding.java:94) ~[?:?]
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
at org.apache.felix.scr.impl.inject.BaseMethod.invokeMethod(BaseMethod.java:229) ~[?:?]
at org.apache.felix.scr.impl.inject.BaseMethod.access$500(BaseMethod.java:39) ~[?:?]
at org.apache.felix.scr.impl.inject.BaseMethod$Resolved.invoke(BaseMethod.java:650) ~[?:?]
at org.apache.felix.scr.impl.inject.BaseMethod.invoke(BaseMethod.java:506) ~[?:?]
at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:307) ~[?:?]
at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:299) ~[?:?]
at org.apache.felix.scr.impl.manager.SingleComponentManager.disposeImplementationObject(SingleComponentManager.java:344) ~[?:?]
at org.apache.felix.scr.impl.manager.SingleComponentManager.deleteComponent(SingleComponentManager.java:153) ~[?:?]
at org.apache.felix.scr.impl.manager.AbstractComponentManager.doDeactivate(AbstractComponentManager.java:814) ~[?:?]
at org.apache.felix.scr.impl.manager.AbstractComponentManager.deactivateInternal(AbstractComponentManager.java:788) ~[?:?]
at org.apache.felix.scr.impl.manager.AbstractComponentManager.dispose(AbstractComponentManager.java:580) ~[?:?]
at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.disposeComponents(ConfigurableComponentHolder.java:706) ~[?:?]
at org.apache.felix.scr.impl.BundleComponentActivator.dispose(BundleComponentActivator.java:523) ~[?:?]
at org.apache.felix.scr.impl.Activator.disposeComponents(Activator.java:439) ~[?:?]
at org.apache.felix.scr.impl.Activator.access$300(Activator.java:54) ~[?:?]
at org.apache.felix.scr.impl.Activator$ScrExtension.destroy(Activator.java:293) ~[?:?]
at org.apache.felix.utils.extender.AbstractExtender$2.run(AbstractExtender.java:285) ~[?:?]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:?]
at org.apache.felix.utils.extender.AbstractExtender.destroyExtension(AbstractExtender.java:307) ~[?:?]
at org.apache.felix.utils.extender.AbstractExtender.bundleChanged(AbstractExtender.java:181) ~[?:?]
at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:903) ~[?:?]
at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) ~[?:?]
at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) ~[?:?]
at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:213) ~[?:?]
at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:120) ~[?:?]
at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:112) ~[?:?]
at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:156) ~[?:?]
at org.eclipse.osgi.container.Module.publishEvent(Module.java:476) ~[?:?]
at org.eclipse.osgi.container.Module.doStop(Module.java:634) ~[?:?]
at org.eclipse.osgi.container.Module.stop(Module.java:498) ~[?:?]
at org.eclipse.osgi.internal.framework.EquinoxBundle.stop(EquinoxBundle.java:410) ~[?:?]
at org.apache.karaf.features.internal.service.FeaturesServiceImpl.stopBundle(FeaturesServiceImpl.java:1367) ~[?:?]
at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:642) ~[?:?]
at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1248) ~[?:?]
at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$1(FeaturesServiceImpl.java:1147) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
at java.lang.Thread.run(Thread.java:748) [?:?]
2017-12-08 20:42:31.040 [INFO ] [core.karaf.internal.FeatureInstaller] - Uninstalled 'openhab-binding-lgtv1'`

## Possible Solution
None. OH2 deinstalls the binding automatically after the error

## Steps to Reproduce (for bugs)
1. Just boot & configure the LG-TV binding corretly

## Context
Nothing special

## Your Environment
openHAB-Version:  openHAB 2.2.0~20171208181417-1 (Build #1119) [openHABian], expierenced the same problems unter 2.1.0
Hardware: Raspberry PI 2

Here some maybe helpful files:

**lgtv.cfg:**

> lgtvwz.host=192.168.1.xx
> lgtvwz.port=8080
> lgtvwz.pairkey=123456
> lgtvwz.serverport=8080
> lgtvwz.xmldatafiles=./
> lgtvwz.checkalive=60
> 

**items definition:**
> String LGTVWZ_AppStart          "Start"                          {lgtv="\*:lgtvwz:APP_EXECUTE"}
> String LGTVWZ_AppStopp          "Stopp"                          {lgtv="\*:lgtvwz:APP_TERMINATE"}
> String LGTVWZ_Status            "Status [%s]"                    {lgtv="\*:lgtvwz:CONNECTION_STATUS"}

**sitemap:**

> Frame label="Tv"
>     {
>         Switch item=LGTVWZ_Status
>         Switch item=LGTVWZ_AppStart mappings=[YouTube=YouTube, LOVEFiLM=Prime]
>         Switch item=LGTVWZ_AppStopp mappings=[YouTube=YouTube, LOVEFiLM=Prime]
>     }
> 

Thanks, Boby

--- a/bundles/binding/org.openhab.binding.lgtv/src/main/java/org/openhab/binding/lgtv/lginteraction/LgTvMessageReader.java
+++ b/bundles/binding/org.openhab.binding.lgtv/src/main/java/org/openhab/binding/lgtv/lginteraction/LgTvMessageReader.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2016 by the respective copyright holders.
+ * Copyright (c) 2010-2017 by the respective copyright holders.
  *
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
@@ -52,7 +52,7 @@ import org.slf4j.LoggerFactory;
  */
 public class LgTvMessageReader extends HttpServlet {
 
-    private static Logger logger = LoggerFactory.getLogger(LgtvConnection.class);
+    private static Logger logger = LoggerFactory.getLogger(LgTvMessageReader.class);
     private static List<LgtvEventListener> _listeners = new ArrayList<LgtvEventListener>();
     protected static ItemRegistry itemRegistry; // mf11102014
     public static final String WEBAPP_ALIAS = "/";
@@ -88,7 +88,7 @@ public class LgTvMessageReader extends HttpServlet {
     protected HttpService httpService;
 
     public void setHttpService(HttpService httpService) {
-        logger.info("sethttpservice called");
+        logger.info("setHttpService called");
         this.httpService = httpService;
     }
 
@@ -125,8 +125,8 @@ public class LgTvMessageReader extends HttpServlet {
      * @return a {@link SecureHttpContext}
      */
     protected HttpContext createHttpContext() {
-        if (this.httpService == null) {
-            logger.error("cannot create http context httpservice null");
+        if (httpService == null) {
+            logger.warn("Cannot create HttpContext because the httpService is null");
             return null;
         } else {
             HttpContext defaultHttpContext = httpService.createDefaultHttpContext();
@@ -140,6 +140,8 @@ public class LgTvMessageReader extends HttpServlet {
     }
 
     public void activate() {
+        logger.info("LgTv servlet activate called");
+
         try {
             bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
             if (bundleContext != null) {
@@ -148,39 +150,42 @@ public class LgTvMessageReader extends HttpServlet {
                 if (serviceReference1 != null) {
                     bindingprovider = (LgtvBindingProvider) bundleContext.getService(serviceReference1);
                 } else {
-                    logger.error("lgtvbindungprovider=null");
+                    logger.warn("Unable to get service reference from LgTv binding provider class");
                 }
 
                 ServiceReference<?> serviceReference2 = bundleContext.getServiceReference(ItemRegistry.class.getName());
                 if (serviceReference2 != null) {
                     itemregistry = (ItemRegistry) bundleContext.getService(serviceReference2);
                 } else {
-                    logger.error("itemregistry=null");
+                    logger.warn("Unable to get service reference from ItemRegistry class");
                 }
             } else {
-                logger.error("bundleContext=null");
+                logger.warn("bundleContext is null. Unable to get binding provider or item registry.");
             }
 
-            logger.info("lgtv servlet activate called");
             Hashtable<String, String> props = new Hashtable<String, String>();
             httpService.registerServlet(WEBAPP_ALIAS + SERVLET_NAME, this, props, createHttpContext());
-            logger.info("Started LgTv Servlet at " + WEBAPP_ALIAS + SERVLET_NAME);
+            logger.info("Started LgTv Servlet at {} {}", WEBAPP_ALIAS, SERVLET_NAME);
         } catch (NamespaceException e) {
-            logger.error("Error during servlet startup", e);
+            logger.warn("NamespaceException occurred during servlet startup", e);
         } catch (ServletException e) {
-            logger.error("Error during servlet startup", e);
+            logger.warn("ServletException occurred during servlet startup", e);
         } catch (Exception e) {
-            logger.error("error during servlet startup", e);
+            logger.warn("Unexpected exception occurred during servlet startup", e);
         }
     }
 
     public void deactivate() {
-        httpService.unregister(WEBAPP_ALIAS + SERVLET_NAME);
-        logger.info("Stopped LgTv Servlet ");
+        try {
+            httpService.unregister(WEBAPP_ALIAS + SERVLET_NAME);
+        } catch(IllegalArgumentException e) {
+            logger.info("LgTv Servlet '{}' was not registered. Nothing to deactivate.", WEBAPP_ALIAS + SERVLET_NAME);
+        } finally {
+            logger.info("Stopped LgTv Servlet");
+        }
     }
 
     public void startserver() throws IOException {
-
         if (status == 0) {
             activate();
             status = 1;
@@ -191,7 +196,6 @@ public class LgTvMessageReader extends HttpServlet {
     }
 
     public void stopserver() throws IOException {
-
         deactivate();
         logger.debug("LgTvMessageReader Server stopped");
         status = 0;
@@ -199,16 +203,15 @@ public class LgTvMessageReader extends HttpServlet {
 
     public void sendtohandlers(LgtvStatusUpdateEvent event, String remoteaddr, String message) {
         // send message to event listeners
-        logger.debug("sendtohandlers remoteaddr=" + remoteaddr + " message=" + message);
+        logger.debug("sendtohandlers remoteaddr={} message={}", remoteaddr, message);
         try {
             Iterator<LgtvEventListener> iterator = _listeners.iterator();
 
             while (iterator.hasNext()) {
                 iterator.next().statusUpdateReceived(event, remoteaddr, message);
             }
-
         } catch (Exception e) {
-            logger.error("Cannot send to EventListeners / maybe not initialized yet", e);
+            logger.warn("Cannot send to EventListeners. Maybe not initialized yet", e);
         }
 
     }
@@ -261,14 +264,14 @@ public class LgTvMessageReader extends HttpServlet {
                                     out.print(va);
 
                                 } catch (ItemNotFoundException e) {
-                                    logger.error("item not found");
+                                    logger.warn("Item not found");
                                 }
                             }
                         }
                     }
 
                 } else {
-                    logger.error("itemregistry=null or bindingprovider=null");
+                    logger.warn("itemregistry or bindingprovider is null");
                 }
 
             } else {
@@ -351,7 +354,7 @@ public class LgTvMessageReader extends HttpServlet {
         }
         remoteaddr = t;
 
-        logger.debug("httphandler called from remoteaddr=" + remoteaddr + " result=" + sb.toString());
+        logger.debug("httphandler called from remoteaddr={} result={}", remoteaddr, sb);
 
         LgTvEventChannelChanged myevent = new LgTvEventChannelChanged();
 
@@ -359,16 +362,15 @@ public class LgTvMessageReader extends HttpServlet {
         try {
             result = myevent.readevent(sb.toString());
         } catch (JAXBException e) {
-            logger.error("error in httphandler", e);
+            logger.warn("Error occurred in httphandler", e);
         }
-        logger.debug("eventresult=" + result);
+        logger.debug("eventresult={}", result);
 
         LgTvEventChannelChanged.envelope envel = myevent.getenvel();
 
         String eventname = envel.getchannel().geteventname();
 
         if (eventname.equals("ChannelChanged")) {
-
             String name = "CHANNEL_CURRENTNAME=" + envel.getchannel().getchname();
             String number = "CHANNEL_CURRENTNUMBER=" + envel.getchannel().getmajor();
             String set = "CHANNEL_SET=" + envel.getchannel().getmajor();
@@ -376,16 +378,12 @@ public class LgTvMessageReader extends HttpServlet {
             sendtohandlers(event, remoteaddr, name);
             sendtohandlers(event, remoteaddr, number);
             sendtohandlers(event, remoteaddr, set);
-
         } else if (eventname.equals("byebye")) {
-
             sendtohandlers(event, remoteaddr, "BYEBYE_SEEN=1");
-
         } else {
-            logger.debug("warning - unhandled event");
+            logger.debug("unhandled event: {}", eventname);
         }
 
         responseBody.close();
     }
-
 }

--- a/bundles/binding/org.openhab.binding.lgtv/src/main/java/org/openhab/binding/lgtv/lginteraction/LgTvMessageReader.java
+++ b/bundles/binding/org.openhab.binding.lgtv/src/main/java/org/openhab/binding/lgtv/lginteraction/LgTvMessageReader.java
@@ -88,7 +88,7 @@ public class LgTvMessageReader extends HttpServlet {
     protected HttpService httpService;
 
     public void setHttpService(HttpService httpService) {
-        logger.info("setHttpService called");
+        logger.debug("setHttpService called");
         this.httpService = httpService;
     }
 
@@ -140,7 +140,7 @@ public class LgTvMessageReader extends HttpServlet {
     }
 
     public void activate() {
-        logger.info("LgTv servlet activate called");
+        logger.debug("LgTv servlet activate called");
 
         try {
             bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
@@ -165,7 +165,7 @@ public class LgTvMessageReader extends HttpServlet {
 
             Hashtable<String, String> props = new Hashtable<String, String>();
             httpService.registerServlet(WEBAPP_ALIAS + SERVLET_NAME, this, props, createHttpContext());
-            logger.info("Started LgTv Servlet at {} {}", WEBAPP_ALIAS, SERVLET_NAME);
+            logger.info("Started LgTv Servlet at {}{}", WEBAPP_ALIAS, SERVLET_NAME);
         } catch (NamespaceException e) {
             logger.warn("NamespaceException occurred during servlet startup", e);
         } catch (ServletException e) {
@@ -179,7 +179,7 @@ public class LgTvMessageReader extends HttpServlet {
         try {
             httpService.unregister(WEBAPP_ALIAS + SERVLET_NAME);
         } catch(IllegalArgumentException e) {
-            logger.info("LgTv Servlet '{}' was not registered. Nothing to deactivate.", WEBAPP_ALIAS + SERVLET_NAME);
+            logger.debug("LgTv Servlet '{}' was not registered. Nothing to deactivate.", WEBAPP_ALIAS + SERVLET_NAME);
         } finally {
             logger.info("Stopped LgTv Servlet");
         }

